src/services/api.ts

```diff
@@ -0,0 +1,205 @@
+  executedSellAmount: string;
+}
+interface GetFeeResponse {
+  amount: string;
+  expirationDate: Date;
+}
+interface EstimateAmountResponse {
+  amount: string;
+  token: string;
+}
+export interface ApiError {
+  errorType: string;
+  description: string;
+}
+interface CallError extends Error {
+  apiError?: ApiError;
+}
+
+function apiKind(kind: OrderKind): string {
+  switch (kind) {
+    case OrderKind.SELL:
+      return "sell";
+    case OrderKind.BUY:
+      return "buy";
+    default:
+      throw new Error(`Unsupported kind ${kind}`);
+  }
+}
+
+function apiSigningScheme(scheme: SigningScheme): string {
+  switch (scheme) {
+    case SigningScheme.EIP712:
+      return "eip712";
+    case SigningScheme.ETHSIGN:
+      return "ethsign";
+    case SigningScheme.EIP1271:
+      return "eip1271";
+    case SigningScheme.PRESIGN:
+      return "presign";
+    default:
+      throw new Error(`Unsupported signing scheme ${scheme}`);
+  }
+}
+
+async function call<T>(
+  route: string,
+  network: string,
+  init?: RequestInit,
+): Promise<T> {
+  const url = `https://protocol-${network}.dev.gnosisdev.com/api/v1/${route}`;
+  const response = await fetch(url, init);
+  const body = await response.text();
+  if (!response.ok) {
+    const error: CallError = new Error(
+      `Calling "${url} ${JSON.stringify(init)} failed with ${
+        response.status
+      }: ${body}`,
+    );
+    try {
+      error.apiError = JSON.parse(body);
+    } catch {
+      // no api error
+    }
+    throw error;
+  }
+  return JSON.parse(body);
+}
+
+export async function getFee({
+  sellToken,
+  buyToken,
+  kind,
+  amount,
+  network,
+}: GetFeeQuery & ApiCall): Promise<BigNumber> {
+  const response: GetFeeResponse = await call(
+    `fee?sellToken=${sellToken}&buyToken=${buyToken}&amount=${amount}&kind=${apiKind(
+      kind,
+    )}`,
+    network,
+  );
+  return BigNumber.from(response.amount);
+}
+
+export async function estimateTradeAmount({
+  network,
+  sellToken,
+  buyToken,
+  kind,
+  amount,
+}: EstimateTradeAmountQuery & ApiCall): Promise<BigNumber> {
+  const response: EstimateAmountResponse = await call(
+    `markets/${sellToken}-${buyToken}/${apiKind(kind)}/${amount}`,
+    network,
+  );
+  // The services return the quote token used for the price. The quote token
+  // is checked to make sure that the returned price meets our expectations.
+  if (response.token.toLowerCase() !== buyToken.toLowerCase()) {
+    throw new Error(
+      `Price returned for sell token ${sellToken} uses an incorrect quote token (${response.token.toLowerCase()} instead of ${buyToken.toLowerCase()})`,
+    );
+  }
+  return BigNumber.from(response.amount);
+}
+
+export async function placeOrder({
+  order,
+  signature,
+  network,
+}: PlaceOrderQuery & ApiCall): Promise<string> {
+  const normalizedOrder = normalizeOrder(order);
+  // the api expects appData to always have length 32 bytes
+  const appData = utils.hexlify(
```

Comment:
nit: maybe this is something we should also enforce on the Order type directly (or is there a reason we use a string there? Will a hex string map correctly to the hex representation or will e.g `0xfeedface` be ASCII encoded?

Reply:
I assumed that any bitstring smaller than 32 bytes would be padded by Hardhat to 32 bytes, but I was mistaken:
```
> var x = new ethers.utils.Interface(["function f(bytes32)"])
> x.encodeFunctionData("f",["0x" + "42".repeat(32)])
'0xd7da973a4242424242424242424242424242424242424242424242424242424242424242'
> x.encodeFunctionData("f",["0x"])
Uncaught:
Error: incorrect data length (argument=null, value="0x", code=INVALID_ARGUMENT, version=abi/5.1.2)
```
So the length is already enforced in the `Order` type. Then I'll remove the padding.
About the hex representation, Hardhat supports multiple representations of a bitstring, one of which is a string of the form `0x...`. `utils.hexlify()` parses any of these representations into a hex string, so `0xfeedface` would be unchanged (also, zero padding is not stripped).